### PR TITLE
Feature array shuffle

### PR DESCRIPTION
**Pull request by jh0x**
Part of #11125 -> introduce `arrayShuffle` function.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce arrayShuffle function for random array permutations.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by jh0x** [comment]
I intend to add the other random functions when the logic here is deemed OK.

File: tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference [hunk 1]
```diff
@@ -130,6 +130,7 @@ arrayReverse
 arrayReverseFill
 arrayReverseSort
 arrayReverseSplit
+arrayShuffle
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:133]: Simply copy-paste the markdown docs content to the documentation object inside the registerFunction.

The `asin.cpp` as an example.

File: src/Functions/array/arrayShuffle.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        size_t count = off - prev_off;
+
+        permutation.resize(count);
+        for (size_t idx = 0; idx < count; ++idx)
+            permutation[idx] = idx;
+        std::shuffle(std::begin(permutation), std::end(permutation), rng);
+        for (size_t unshuffled_idx = 0; unshuffled_idx != count; ++unshuffled_idx)
+        {
+            auto shuffled_idx = permutation[unshuffled_idx];
+            res_data.insertFrom(src_data, shuffled_idx);
```
Comment (alexey-milovidov) [inline src/Functions/array/arrayShuffle.cpp:155]: Can we use the IColumn::permute method?

Comment (alexey-milovidov) [reply src/Functions/array/arrayShuffle.cpp:155]: So, we will create a big permutation for all the arrays' data, but shuffle this permutation only inside each array, and then use it to permute the data.

This method might appear to be as efficient as specializations for various types.

Comment (jh0x) [reply src/Functions/array/arrayShuffle.cpp]: Should all types be handled in a generic way (`IColumn::permute`) - as in the latest commit - or choose some mid-ground -> e.g. specialised numbers + rest generic?

Comment (jh0x) [reply src/Functions/array/arrayShuffle.cpp]: The only difference would be that for numbers we don't need to create the permutations array. Instead we `memcpy` the data and shuffle directly in the result (+ optionally the null column). 
Previous logic can be restored by reverting: [FIXUP: Simplify logic by using permute function](https://github.com/ClickHouse/ClickHouse/pull/45271/commits/cf1c5e4ac0b506789d7744a34acc24d3d06bfe6f)

**Comment by jh0x** [comment]
I added implementation for `arrayPartialShuffle`.
I also have a question about the documentation - should it be just in code? Or should I duplicate it in `docs` markdown.

File: src/Functions/array/arrayShuffle.cpp [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                {"explicit_seed", "SELECT arrayShuffle([1, 2, 3, 4], 41)"}},
+            Documentation::Categories{"Array"}
+        },
+        FunctionFactory::CaseInsensitive);
+    factory.registerFunction<FunctionArrayShuffleImpl<FunctionArrayPartialShuffleTraits>>(
+        {
+            R"(
+Returns an array of the same size as the original array where elements in range [0..limit) are a random
+subset of the original array. Remaining [limit..n) shall contain the elements not in [0..limit) range in undefined order.
+Value of limit shall be in range [0..n]. Values outside of that range are equivalent to performing full arrayShuffle:
```
Comment (vdimir) [inline src/Functions/array/arrayShuffle.cpp:198]: It's ok  eitgher to keed doc only in cpp and have duplicating md. One in cpp is source of truth.

---

What the purpose of this function? Is it similar to `arrayConcat(arrayShuffle(arraySlice(a, 1, limit)), arraySlice(a, limit+1))` but with no guaranteed order for tail?

---

Also array indexing in ClickHouse are 1-based, so description can be adjusted

File: tests/queries/0_stateless/02523_array_shuffle.sql [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+SELECT arrayShuffle(['storage','tiger','imposter','terminal','uniform','sensation'], 0xbad_cafe);
+SELECT arrayShuffle(['storage','tiger',NULL,'terminal','uniform','sensation'], 0xbad_cafe);
+SELECT arrayShuffle([NULL]);
+SELECT arrayShuffle([NULL,NULL]);
+SELECT arrayShuffle([[1,2,3,4],[-1,-2,-3,-4],[10,20,30,40],[100,200,300,400,500,600,700,800,900],[2,4,8,16,32,64]], 0xbad_cafe);
+SELECT arrayShuffle([[1,2,3,4],[NULL,-2,-3,-4],[10,20,30,40],[100,200,300,400,500,600,700,800,900],[2,4,8,16,32,64]], 0xbad_cafe);
+SELECT arrayShuffle(groupArray(x),0xbad_cafe) FROM (SELECT number as x from system.numbers LIMIT 100);
+SELECT arrayShuffle(groupArray(toUInt64(x)),0xbad_cafe) FROM (SELECT number as x from system.numbers LIMIT 100);
+SELECT arrayShuffle([tuple(1, -1), tuple(99999999, -99999999), tuple(3, -3)], 0xbad_cafe);
+SELECT arrayShuffle([tuple(1, NULL), tuple(2, 'a'), tuple(3, 'A')], 0xbad_cafe);
```
Comment (vdimir) [inline tests/queries/0_stateless/02523_array_shuffle.sql:20]: We don't materialize constant in this functions, so result is different for constant and materialized column:

```sql

SELECT
    arrayShuffle(materialize([1, 2, 3]), 42),
    arrayShuffle([1, 2, 3], 42)
FROM numbers(10)

Query id: 6dfe07fe-b4b5-4139-a539-9914c6e1abf5

┌─arrayShuffle(materialize([1, 2, 3]), 42)─┬─arrayShuffle([1, 2, 3], 42)─┐
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,1,2]                                  │ [3,2,1]                     │
│ [1,3,2]                                  │ [3,2,1]                     │
│ [2,1,3]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [1,2,3]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [2,1,3]                                  │ [3,2,1]                     │
└──────────────────────────────────────────┴─────────────────────────────┘
```

This behavior is ok, but should tested as well (and probably documented)

**Review by vdimir — COMMENTED** [review]
Sorry for the delay, couple of comments more

File: tests/queries/0_stateless/02523_array_shuffle.sql (hunk 4 above)
Comment (jh0x) [reply tests/queries/0_stateless/02523_array_shuffle.sql:20]: Good point. `seed` is being used once for the whole data to be shuffled. We could reseed after each block, **but not sure if this is required (or even intended)**:

```diff
diff --git a/src/Functions/array/arrayShuffle.cpp b/src/Functions/array/arrayShuffle.cpp
index 6532fb89ae..752d4a0ccc 100644
--- a/src/Functions/array/arrayShuffle.cpp
+++ b/src/Functions/array/arrayShuffle.cpp
@@ -102,7 +102,7 @@ public:
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t) const override;
 
 private:
-    static ColumnPtr executeGeneric(const ColumnArray & array, pcg64_fast & rng, size_t limit);
+    static ColumnPtr executeGeneric(const ColumnArray & array, uint64_t seed, size_t limit);
 };
 
 template <typename Traits>
@@ -121,7 +121,6 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeImpl(const ColumnsWithTypeAnd
         const auto * val = arguments[Traits::seed_param_idx].column.get();
         return val->getUInt(0);
     }();
-    pcg64_fast rng(seed);
 
     size_t limit = [&]() -> size_t
     {
@@ -136,14 +135,16 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeImpl(const ColumnsWithTypeAnd
         return 0;
     }();
 
-    return executeGeneric(*array, rng, limit);
+    return executeGeneric(*array, seed, limit);
 }
 
 template <typename Traits>
-ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & array, pcg64_fast & rng, size_t limit [[maybe_unused]])
+ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & array, uint64_t seed, size_t limit [[maybe_unused]])
 {
     const ColumnArray::Offsets & offsets = array.getOffsets();
 
+    pcg64_fast rng(seed);
+
     size_t size = offsets.size();
     size_t nested_size = array.getData().size();
     IColumn::Permutation permutation(nested_size);
@@ -166,6 +167,7 @@ ColumnPtr FunctionArrayShuffleImpl<Traits>::executeGeneric(const ColumnArray & a
         else
             shuffle(&permutation[current_offset], &permutation[next_offset], rng);
         current_offset = next_offset;
+        rng.seed(seed);
     }
     return ColumnArray::create(array.getData().permute(permutation, 0), array.getOffsetsPtr());
 }
```

This results in:
```sql
SELECT
    arrayShuffle(materialize([1, 2, 3]), 42),
    arrayShuffle([1, 2, 3], 42)
FROM numbers(10)

Query id: 4a666b06-ecc8-4c30-b8c3-82e3d1fde4ae

┌─arrayShuffle(materialize([1, 2, 3]), 42)─┬─arrayShuffle([1, 2, 3], 42)─┐
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
│ [3,2,1]                                  │ [3,2,1]                     │
└──────────────────────────────────────────┴─────────────────────────────
```

File: src/Functions/array/arrayShuffle.cpp (hunk 3 above)
Comment (jh0x) [reply src/Functions/array/arrayShuffle.cpp:198]: As for the shuffling logic - we do the following.


For the full shuffle:
```
for i∈[0, n-2):
	j ← random from ∈ [i, n)
	swap arr[i] ↔ arr[j]
```

And for the partial shuffle we do almost exactly the same:
```
for i∈[0, limit):				// Swap only "limit" number of times
	j ← random from ∈ [i, n)		// But swap using the whole array contents
	swap arr[i] ↔ arr[j]
```
Partial shuffle is like [the one in D](https://dlang.org/library/std/random/partial_shuffle.html). We get `limit` elements ~drawn from the whole array.

This is different from the following - here we permute only a small range of elements with themselves:
```sql
SELECT
    groupArray(x) AS arr,
    3 AS lim,
    arrayConcat(arrayShuffle(arraySlice(arr, 1, lim), 41), arraySlice(arr, lim + 1)) AS subsetShuffle,
    arrayPartialShuffle(arr, lim, 41) AS partialShuffle
FROM
(
    SELECT number AS x
    FROM system.numbers
    LIMIT 10
)

Query id: 79b13717-3358-479b-96c5-a6d2c02463f0

┌─arr───────────────────┬─lim─┬─subsetShuffle─────────┬─partialShuffle────────┐
│ [0,1,2,3,4,5,6,7,8,9] │   3 │ [2,1,0,3,4,5,6,7,8,9] │ [4,8,6,3,0,5,2,7,1,9] │
└───────────────────────┴─────┴───────────────────────┴───────────────────────┘
```

---

I'll adjust the doc about indexing.

File: tests/queries/0_stateless/02523_array_shuffle.sql (hunk 4 above)
Comment (vdimir) [reply tests/queries/0_stateless/02523_array_shuffle.sql:20]: `arrayShuffle(materialize([1, 2, 3]), 42)` should return different permutations for each row. I meant, for constant array, we can do the same (materialize it automatically), but probably current behavior is even better.

File: src/Functions/array/arrayShuffle.cpp (hunk 3 above)
Comment (vdimir) [reply src/Functions/array/arrayShuffle.cpp:198]: Got it, thanks for the explanation

**Review by vdimir — APPROVED** [review]
Thanks for your efforts!

**Pull request merged**